File: cpp/fx-options-engine/tests/test_binomial.cpp
// CRR tree: convergence to GK, American vs European ordering, and the
// FX-specific early-exercise economics (foreign carry as 'dividend').

#include <gtest/gtest.h>

#include <cmath>
#include <numeric>
#include <stdexcept>
#include <vector>

#include "fxopt/binomial.hpp"
#include "fxopt/garman_kohlhagen.hpp"

namespace {

using namespace fxopt;

constexpr double S = 1.10, K = 1.08, T = 0.75, RD = 0.0425, RF = 0.0290,
                 SIG = 0.0925;

TEST(Binomial, EuropeanConvergesToGarmanKohlhagen) {
    const double analytic = gk_call(S, K, T, RD, RF, SIG);
    double prev_err = 1e9;
    for (const int steps : {50, 200, 800}) {
        const double tree = binomial_price(S, K, T, RD, RF, SIG,
                                           OptionType::Call, steps);
        const double err = std::abs(tree - analytic);
        EXPECT_LT(err, prev_err * 1.05);  // roughly monotone O(1/n) decay
        prev_err = err;
    }
    EXPECT_NEAR(binomial_price(S, K, T, RD, RF, SIG, OptionType::Call, 2000),
                analytic, 5e-6);
    EXPECT_NEAR(binomial_price(S, K, T, RD, RF, SIG, OptionType::Put, 2000),
                gk_put(S, K, T, RD, RF, SIG), 5e-6);
}

TEST(Binomial, ConvergenceRateFittedExponentMatchesTheoreticalOrderOne) {
    // CRR is a first-order scheme: error(n) ~ C / n, with an odd/even (and
    // node-alignment) oscillation superposed. A two-point ratio (as in
    // EuropeanConvergesToGarmanKohlhagen above) cannot *prove* the O(1/n)
    // rate -- it can land anywhere in the oscillation. Regressing
    // log|error| on log(n) over a decade of geometrically spaced step
    // counts averages the oscillation out and recovers the leading
    // exponent, which must come out close to the theoretical -1.
    const double analytic = gk_call(S, K, T, RD, RF, SIG);
    const std::vector<int> steps{200, 400, 800, 1600, 3200, 6400, 12800, 25600, 51200};
    std::vector<double> log_n, log_err;
    for (int n : steps) {
        const double err = std::abs(
            binomial_price(S, K, T, RD, RF, SIG, OptionType::Call, n) -
            analytic);
        ASSERT_GT(err, 0.0) << "tree exactly matches GK at n=" << n;
        log_n.push_back(std::log(static_cast<double>(n)));
        log_err.push_back(std::log(err));
    }
    const double mean_x =
        std::accumulate(log_n.begin(), log_n.end(), 0.0) / log_n.size();
    const double mean_y =
        std::accumulate(log_err.begin(), log_err.end(), 0.0) / log_err.size();
    double num = 0.0, den = 0.0;
    for (size_t i = 0; i < log_n.size(); ++i) {
        num += (log_n[i] - mean_x) * (log_err[i] - mean_y);
        den += (log_n[i] - mean_x) * (log_n[i] - mean_x);
    }
    const double slope = num / den;
    EXPECT_GT(slope, -1.3) << "fitted CRR exponent " << slope;
    EXPECT_LT(slope, -0.7) << "fitted CRR exponent " << slope;
}

TEST(Binomial, AmericanAtLeastEuropean) {
    for (const auto type : {OptionType::Call, OptionType::Put})
        for (const double k : {0.95, 1.10, 1.25}) {
            const double eur = binomial_price(S, k, T, RD, RF, SIG, type, 400,
                                              Exercise::European);
            const double amer = binomial_price(S, k, T, RD, RF, SIG, type,
                                               400, Exercise::American);
            EXPECT_GE(amer, eur - 1e-12);
        }
}

TEST(Binomial, EarlyExercisePremiumWhenForeignRateHigh) {
    // Economic test: an American call on a high-yielding foreign currency
    // (r_f > r_d) forfeits foreign carry while unexercised -> strictly
    // positive early-exercise premium for an ITM call.  Mirrors the
    // dividend-yield story for equity calls.
    const double rd = 0.005, rf = 0.0525;  // e.g. JPY investor holding USD
    const double s = 147.5, k = 140.0, t = 1.0, sig = 0.105;
    const double eur = binomial_price(s, k, t, rd, rf, sig, OptionType::Call,
                                      800, Exercise::European);
    const double amer = binomial_price(s, k, t, rd, rf, sig, OptionType::Call,
                                       800, Exercise::American);
    EXPECT_GT(amer - eur, 1e-3);
    // Control: with r_d > r_f the ITM *put* carries the premium instead.
    const double eur_put = binomial_price(S, 1.20, T, 0.0525, 0.005, SIG,
                                          OptionType::Put, 800,
                                          Exercise::European);
    const double amer_put = binomial_price(S, 1.20, T, 0.0525, 0.005, SIG,
                                           OptionType::Put, 800,
                                           Exercise::American);
    EXPECT_GT(amer_put - eur_put, 1e-5);
}

TEST(Binomial, DegenerateLimits) {
    // 1.10 - 1.00 (not the literal 0.10): match the intrinsic's own
    // floating-point subtraction.
    EXPECT_DOUBLE_EQ(
        binomial_price(1.10, 1.00, 0.0, RD, RF, SIG, OptionType::Call),
        1.10 - 1.00);
    // sigma = 0 European defers to the GK zero-vol limit.
    EXPECT_NEAR(binomial_price(S, K, T, RD, RF, 0.0, OptionType::Call),
                gk_call(S, K, T, RD, RF, 0.0), 1e-15);
    // sigma = 0 American: deterministic drifting spot, best exercise value.
    const double amer0 = binomial_price(S, 0.9, T, 0.005, 0.0525, 0.0,
                                        OptionType::Call, 500,
                                        Exercise::American);
    EXPECT_GE(amer0, S - 0.9 - 1e-12);  // at least immediate exercise
}

TEST(Binomial, ThrowsOnBadInputs) {
    EXPECT_THROW(binomial_price(S, K, T, RD, RF, SIG, OptionType::Call, 0),
                 std::invalid_argument);
    EXPECT_THROW(binomial_price(-1.0, K, T, RD, RF, SIG, OptionType::Call),
                 std::invalid_argument);
    // dt too coarse for drift/vol -> p outside [0,1].
    EXPECT_THROW(binomial_price(S, K, 10.0, 0.50, 0.0, 0.01,
                                OptionType::Call, 1),
                 std::invalid_argument);
}

}  // namespace

File: python/fx/01-options-pricing/tests/test_binomial.py
"""CRR binomial tree: convergence to GK, American exercise economics."""

import numpy as np
import pytest

from fx_options import binomial_convergence, binomial_price, gk_price

MKT = dict(S=1.10, K=1.12, T=0.5, r_d=0.0425, r_f=0.0290, sigma=0.0825)


class TestEuropeanConvergence:
    @pytest.mark.parametrize("option_type", ["call", "put"])
    def test_converges_to_gk(self, option_type):
        tree = binomial_price(**MKT, option_type=option_type, steps=2000)
        gk = gk_price(**MKT, option_type=option_type)
        assert tree == pytest.approx(gk, abs=5e-6)

    def test_error_decreases_with_steps(self):
        rows = binomial_convergence(**MKT, option_type="call",
                                    step_grid=(10, 100, 1000))
        errors = [r["abs_error"] for r in rows]
        assert errors[2] < errors[0]
        assert errors[2] < 1e-5

    def test_convergence_rate_fitted_exponent(self):
        """Fit the CRR discretisation-error exponent by log-log regression.

        CRR is a first-order scheme: error(n) ~ C / n, with an odd/even
        and node-alignment oscillation superposed. A single pair of step
        counts (as in ``test_error_decreases_with_steps`` above) cannot
        *prove* the O(1/n) rate -- a two-point ratio can land anywhere in
        the oscillation. Regressing log|error| on log(n) over a decade of
        geometrically spaced step counts averages the oscillation out and
        recovers the leading exponent, which must come out close to the
        theoretical -1.
        """
        steps = tuple(200 * 2**k for k in range(9))  # 200 .. 51200
        rows = binomial_convergence(**MKT, option_type="call", step_grid=steps)
        errs = np.array([r["abs_error"] for r in rows])
        assert np.all(errs > 0.0), "tree exactly matches GK at some n"
        slope, _ = np.polyfit(np.log(steps), np.log(errs), 1)
        assert -1.3 < slope < -0.7, (
            f"fitted CRR convergence exponent {slope:.3f}; theory predicts "
            "-1 (error ~ C/n)"
        )

    def test_jpy_level_convergence(self):
        # High spot level (pip 0.01) must not degrade accuracy in relative terms.
        args = dict(S=147.5, K=145.0, T=0.5, r_d=0.005, r_f=0.0525,
                    sigma=0.1075)
        tree = binomial_price(**args, option_type="put", steps=2000)
        gk = gk_price(**args, option_type="put")
        assert tree == pytest.approx(gk, rel=3e-4)


class TestAmerican:
    @pytest.mark.parametrize("option_type", ["call", "put"])
    def test_american_geq_european(self, option_type):
        eu = binomial_price(**MKT, option_type=option_type, steps=500,
                            exercise="european")
        am = binomial_price(**MKT, option_type=option_type, steps=500,
                            exercise="american")
        assert am >= eu - 1e-12

    def test_american_call_premium_positive_when_rf_above_rd(self):
        # Foreign carry (r_f > r_d) makes early exercise of a call on the
        # base currency optimal: USDJPY-style. Economically the option
        # holder forgoes the high USD deposit rate while unexercised.
        args = dict(S=147.5, K=140.0, T=1.0, r_d=0.005, r_f=0.0525,
                    sigma=0.1075)
        eu = binomial_price(**args, option_type="call", steps=1000,
                            exercise="european")
        am = binomial_price(**args, option_type="call", steps=1000,
                            exercise="american")
        assert am - eu > 1e-3  # economically significant premium

    def test_american_call_no_premium_when_rf_zero_ish(self):
        # With r_f <= r_d an American FX call is worth its European value
        # (no carry incentive), like a no-dividend equity call.
        eu = binomial_price(**MKT, option_type="call", steps=500,
                            exercise="european")
        am = binomial_price(**MKT, option_type="call", steps=500,
                            exercise="american")
        assert am == pytest.approx(eu, abs=1e-10)

    def test_american_put_premium_when_rd_high(self):
        args = dict(S=1.10, K=1.15, T=1.0, r_d=0.08, r_f=0.00, sigma=0.10)
        eu = binomial_price(**args, option_type="put", steps=1000,
                            exercise="european")
        am = binomial_price(**args, option_type="put", steps=1000,
                            exercise="american")
        assert am - eu > 1e-4

    def test_american_at_least_intrinsic(self):
        am = binomial_price(S=1.30, K=1.10, T=0.5, r_d=0.02, r_f=0.05,
                            sigma=0.10, option_type="call", steps=200,
                            exercise="american")
        assert am >= 0.20 - 1e-12


class TestLimitsAndValidation:
    def test_t_zero_intrinsic(self):
        assert binomial_price(1.2, 1.1, 0.0, 0.03, 0.01, 0.1, "call",
                              steps=10) == pytest.approx(0.1)

    def test_sigma_zero_matches_gk_european(self):
        val = binomial_price(1.1, 1.05, 0.5, 0.04, 0.01, 0.0, "call",
                             steps=100)
        assert val == pytest.approx(
            gk_price(1.1, 1.05, 0.5, 0.04, 0.01, 0.0, "call"), abs=1e-14)

    def test_invalid_steps_raise(self):
        with pytest.raises(ValueError, match="steps"):
            binomial_price(**MKT, option_type="call", steps=0)
        with pytest.raises(ValueError, match="steps"):
            binomial_price(**MKT, option_type="call", steps=2.5)  # type: ignore[arg-type]

    def test_invalid_exercise_raises(self):
        with pytest.raises(ValueError, match="exercise"):
            binomial_price(**MKT, option_type="call", steps=10,
                           exercise="bermudan")

    def test_invalid_market_inputs_raise(self):
        with pytest.raises(ValueError):
            binomial_price(-1.0, 1.1, 0.5, 0.03, 0.01, 0.1, "call")
